nemo/src/model/rule_model/rule.rs

```diff
@@ -11,158 +11,156 @@ pub struct Rule {
-        let existential_variables = body
+        // All the existential variables used in the rule
+        let existential_variable_names = head
             .iter()
-            .flat_map(|literal| literal.existential_variables())
-            .collect::<Vec<_>>();
-
-        if !existential_variables.is_empty() {
-            return Err(ParseError::BodyExistential(
-                existential_variables
-                    .first()
-                    .expect("is not empty here")
-                    .name(),
-            ));
+            .flat_map(|a| a.existential_variables().map(|v| v.name()))
+            .collect::<HashSet<_>>();
+
+        for variable in body
+            .iter()
+            .flat_map(|l| l.variables())
+            .chain(conditions.iter().flat_map(|c| c.variables()))
+        {
+            // Existential variables may only occur in the head
+            if variable.is_existential() {
+                return Err(ParseError::BodyExistential(variable.name()));
+            }
+
+            // There may not be a universal variable whose name is the same that of an existential
+            if existential_variable_names.contains(&variable.name()) {
+                return Err(ParseError::BothQuantifiers(variable.name()));
+            }
         }
 
-        // Check if some variable in the body occurs only in negative literals.
+        // Divide the literals into a positive and a negative part
         let (positive, negative): (Vec<_>, Vec<_>) = body
             .iter()
             .cloned()
             .partition(|literal| literal.is_positive());
-        let positive_varibales = positive
-            .iter()
-            .flat_map(|literal| literal.universal_variables())
-            .collect::<HashSet<&Variable>>();
 
-        let mut unsafe_negative_variables = HashSet::<Variable>::new();
-        for negative_literal in negative {
-            let mut current_unsafe = HashSet::<Variable>::new();
+        // Bound variables are considered to be all variables occuring as primitive terms in a positive body literal
+        let bound_variables = positive
+            .iter()
+            .flat_map(|l| l.primitive_terms())
+            .filter_map(|t| {
+                if let PrimitiveTerm::Variable(variable) = t {
+                    Some(variable)
+                } else {
+                    None
+                }
+            })
+            .collect::<HashSet<_>>();
 
-            for variable in negative_literal.variables() {
-                if positive_varibales.contains(variable) {
-                    continue;
+        // A derived variable is a variable which is defined in terms of bound variables
+        let mut derived_variables = HashSet::<&Variable>::new();
+        for condition in &conditions {
+            if let Condition::Assignment(variable, term) = condition {
+                if bound_variables.contains(variable) || derived_variables.contains(variable) {
```

Comment:
Is it totally unreasonable to have more than one equality constraint on one variable? I guess something like
`p(?prev, ?x, ?next) :- q(?prev), q(?x), q(?next), ?x = ?prev - 1, ?x = ?next + 1.`
might be desirable.

Reply:
Assignment is another operator `:=` which works different from equality. You can have multiple equality restrictions on one variable